Core/Src/ADXL.c

```diff
@@ -10,21 +10,86 @@
 #include "main.h"
 #include <stdbool.h>
 #include <errno.h>
-#include <stm32l4xx_hal_dma.h>
 
 extern SPI_HandleTypeDef hspi1;
-#define ADXL_SPI_ReadOp (1 << 7)
-#define ADXL_SPI_MB     (1 <<6)
-#define ADXL_INT_ENABLE_REG 0x2E
-#define ADXL_INT_MAP_REG    0x2F
-#define ADXL_DEVICEID       0x00
-bool ACC_EXTI_Interrupt;
-static uint8_t TxBuffer[] = { 0, 0, 0, 0, 0, 0, 0 };
-static uint8_t RxBuffer[7];
-
 extern DMA_HandleTypeDef hdma_spi1_rx;
 
-bool ADXL_WriteData(uint8_t regAddress, uint8_t registerValue) {
+#define ADXL_SPI_ReadOp             (1 << 7)
+#define ADXL_SPI_MB                 (1 <<6)
+#define ADXL_INT_ENABLE_REG         0x2E
+#define ADXL_INT_MAP_REG            0x2F
+#define ADXL_DEVICEID               0x00
+#define ADXL_DATAX0                 0x32
+#define ADXL_POWER_CTL              0x2D
+#define ADXL345_SCALE_FACTOR        39  //  if is 2g  , if is 4g *2  8g *4 16g*8
+
+static bool _ReadData(uint8_t regAddress, uint8_t *registerValue);
+static void _Set_Measurebit(void);
+static void _Stop_Measurebit(void);
+static void _Activate_It(void);
+static void _readValuesXYZ(int16_t *x, int16_t *y, int16_t *z);
+
+bool ACC_EXTI_Interrupt_g;
+
+static uint8_t TxBuffer_s[] = { 0, 0, 0, 0, 0, 0, 0 };
+static uint8_t RxBuffer_s[7];
+
+void ADXL_Init(void)
+{
+    uint8_t devId;
+    int16_t x, y, z;
+
+    //reset ADXL
+    HAL_GPIO_TogglePin(ADXL_RESET_GPIO_Port, ADXL_RESET_Pin);
+    HAL_Delay(100);
+    HAL_GPIO_TogglePin(ADXL_RESET_GPIO_Port, ADXL_RESET_Pin);
+
+    // Enter low power mode
+    _Stop_Measurebit();
+
+    // Check SPI Communication
+    _ReadData(ADXL_DEVICEID, &devId);
+    if (devId != 0xE5)
+    {
+        Error_Handler();
+    }
+    _Activate_It();
+
+    // Start Measure
+    _Set_Measurebit();
+
+    //make dummy read to clear first EXTI interrupt
+    _readValuesXYZ(&x, &y, &z);
+}
+void ADXL_GetValues(int16_t *x, int16_t *y, int16_t *z)
+{
+    *x = (int16_t)(RxBuffer_s[2] << 8 | RxBuffer_s[1]);
+    *y = (int16_t)(RxBuffer_s[4] << 8 | RxBuffer_s[3]);
+    *z = (int16_t)(RxBuffer_s[6] << 8 | RxBuffer_s[5]);
+    *x = (*x) * ADXL345_SCALE_FACTOR;
+    *y = (*y) * ADXL345_SCALE_FACTOR;
+    *z = (*z) * ADXL345_SCALE_FACTOR;
+}
+
+void ADXL_StartMeasurement(void)
+{
+    for(int i = 1; i < 7; i++)
+    {
+        TxBuffer_s[i] = 0;
+    }
+
+    TxBuffer_s[0] = ADXL_DATAX0 | ADXL_SPI_MB | ADXL_SPI_ReadOp;
+
+    if(HAL_SPI_GetState(&hspi1) == HAL_SPI_STATE_READY)
+    {
+        ACC_EXTI_Interrupt_g = true;
+        HAL_GPIO_WritePin(SPI_CS_ACC_GPIO_Port, SPI_CS_ACC_Pin, GPIO_PIN_RESET);
+        HAL_SPI_TransmitReceive_DMA(&hspi1, TxBuffer_s, RxBuffer_s, 7);
+    }
+}
+
+static bool _WriteData(uint8_t regAddress, uint8_t registerValue)
```

Comment:
a write function should not return a boolean value.
Usually read/write functions have the following standard:
Returns the number of bytes successfully read.
Returns a negative value if an error has occured.
Returns 0 if at end of file, etc.

Read more here:
https://en.wikipedia.org/wiki/Write_(system_call)